### PR TITLE
Add script to automate everything

### PR DESCRIPTION
**Pull request by bartsch-dev**
Adresses:  #74 & #41
Should help solving: #19

I wrote a script with python which generates a lot and helps maintaining the help files.

it has 3 mode:
- update
  - If a page is missing in a language but present in English it will generate a redirection to the English version. (many examples in my 3rd commit)
  - If a page was renamed and redirecting to the new one in English it only needs to defined in English and they will be generated in all other languages (redirecting to their own language of course, e.g. `en/LabelPatterns.md`)
   - it will generate the index files of each language with all available files
      - in the frontmatter of each help files the categoreis they reside in have to be defined (it suffices if they are only in the English file)
      - The order of the categories is kept in `_scripts/categories.json`
      - The translation of the categories are kept in `_scripts/localization/en.json` while `en` is the language coe

- status
  - collects the current status and prints it on the terminal and creates a markdown file on demand (look below)

- clean 
 - deletes all the generated files created with this script (including the index files)

### When renaming help files
 all but the English one can be deleted and the English one should contain: (where redirect should of course link to the new page)
```
---
redirect:  /en/BibtexKeyPatterns
layout:    redirect
---
```

# status after I execute my script
### Help pages status (2016-09-15 14:14 - Branch `addScriptToAutomateEverything` `0962a3b`)
- Main language: `en`
- Available languages: `de`, `fr`, `in`, `ja`

| Language | translated | not translated | outdated | old pages | % translated | % outdated |
| -------- | ---------- | -------------- | -------- | --------- | ------------ | ---------- |
| de | 32 | 24 | 3 | 1 | 57.1428571429 | 9.375 |
| en | 56 | 0 | 7 | 1 | 100.0 | 12.5 |
| fr | 42 | 14 | 3 | 1 | 75.0 | 7.14285714286 |
| in | 22 | 34 | 1 | 1 | 39.2857142857 | 4.54545454545 |
| ja | 42 | 14 | 3 | 1 | 75.0 | 7.14285714286 |

- Outdated page(s):
  - `de`
    - [CommandLine.md](https://github.com/JabRef/help.jabref.org/blob/gh-pages/de/CommandLine.md)
    - [EntryEditorHelp.md](https://github.com/JabRef/help.jabref.org/blob/gh-pages/de/EntryEditorHelp.md)
    - [FileLinks.md](https://github.com/JabRef/help.jabref.org/blob/gh-pages/de/FileLinks.md)
  - `en`
    - [ContentSelectorHelp.md](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/ContentSelectorHelp.md)
    - [EndNoteFilters.md](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/EndNoteFilters.md)
    - [EntryEditorHelp.md](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/EntryEditorHelp.md)
    - [ExternalFiles.md](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/ExternalFiles.md)
    - [FileLinks.md](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/FileLinks.md)
    - [SearchHelp.md](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/SearchHelp.md)
    - [SpecialFieldsHelp.md](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/SpecialFieldsHelp.md)
  - `fr`
    - [CommandLine.md](https://github.com/JabRef/help.jabref.org/blob/gh-pages/fr/CommandLine.md)
    - [EntryEditorHelp.md](https://github.com/JabRef/help.jabref.org/blob/gh-pages/fr/EntryEditorHelp.md)
    - [FileLinks.md](https://github.com/JabRef/help.jabref.org/blob/gh-pages/fr/FileLinks.md)
  - `in`
    - [EntryEditorHelp.md](https://github.com/JabRef/help.jabref.org/blob/gh-pages/in/EntryEditorHelp.md)
  - `ja`
    - [CommandLine.md](https://github.com/JabRef/help.jabref.org/blob/gh-pages/ja/CommandLine.md)
    - [EntryEditorHelp.md](https://github.com/JabRef/help.jabref.org/blob/gh-pages/ja/EntryEditorHelp.md)
    - [FileLinks.md](https://github.com/JabRef/help.jabref.org/blob/gh-pages/ja/FileLinks.md)
- Not translated page(s):
  - `de`
    - ACMPortalHelp.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/ACMPortalHelp.md))
    - ADSHelp.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/ADSHelp.md))
    - arXivHelp.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/arXivHelp.md))
    - Autosave.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/Autosave.md))
    - BestPractices.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/BestPractices.md))
    - DatabaseProperties.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/DatabaseProperties.md))
    - DBLPHelp.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/DBLPHelp.md))
    - DiVAtoBibTeXHelp.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/DiVAtoBibTeXHelp.md))
    - DOAJHelp.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/DOAJHelp.md))
    - DOItoBibTeXHelp.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/DOItoBibTeXHelp.md))
    - FindDuplicates.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/FindDuplicates.md))
    - GoogleScholarHelp.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/GoogleScholarHelp.md))
    - GVKHelp.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/GVKHelp.md))
    - Installation.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/Installation.md))
    - ISBNtoBibTeXHelp.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/ISBNtoBibTeXHelp.md))
    - MsOfficeBibFieldMapping.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/MsOfficeBibFieldMapping.md))
    - ScienceDirect.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/ScienceDirect.md))
    - SharedBibFile.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/SharedBibFile.md))
    - SpecialFieldsHelp.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/SpecialFieldsHelp.md))
    - SpringerHelp.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/SpringerHelp.md))
    - SQLDatabase.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/SQLDatabase.md))
    - SQLDatabaseMigration.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/SQLDatabaseMigration.md))
    - SQLExport.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/SQLExport.md))
    - StringsHelp.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/StringsHelp.md))
  - `fr`
    - ADSHelp.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/ADSHelp.md))
    - arXivHelp.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/arXivHelp.md))
    - BestPractices.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/BestPractices.md))
    - DBLPHelp.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/DBLPHelp.md))
    - DOAJHelp.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/DOAJHelp.md))
    - FindDuplicates.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/FindDuplicates.md))
    - GVKHelp.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/GVKHelp.md))
    - Installation.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/Installation.md))
    - MedlineRIS.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/MedlineRIS.md))
    - MsOfficeBibFieldMapping.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/MsOfficeBibFieldMapping.md))
    - SharedBibFile.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/SharedBibFile.md))
    - SpringerHelp.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/SpringerHelp.md))
    - SQLDatabase.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/SQLDatabase.md))
    - SQLDatabaseMigration.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/SQLDatabaseMigration.md))
  - `in`
    - ADSHelp.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/ADSHelp.md))
    - arXivHelp.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/arXivHelp.md))
    - BestPractices.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/BestPractices.md))
    - BibtexKeyPatterns.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/BibtexKeyPatterns.md))
    - CommandLine.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/CommandLine.md))
    - CustomEntriesHelp.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/CustomEntriesHelp.md))
    - CustomImports.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/CustomImports.md))
    - DBLPHelp.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/DBLPHelp.md))
    - DiVAtoBibTeXHelp.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/DiVAtoBibTeXHelp.md))
    - DOAJHelp.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/DOAJHelp.md))
    - DOItoBibTeXHelp.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/DOItoBibTeXHelp.md))
    - EndNoteFilters.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/EndNoteFilters.md))
    - ExternalFiles.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/ExternalFiles.md))
    - FileLinks.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/FileLinks.md))
    - FindDuplicates.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/FindDuplicates.md))
    - GeneralFields.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/GeneralFields.md))
    - GoogleScholarHelp.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/GoogleScholarHelp.md))
    - GroupsHelp.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/GroupsHelp.md))
    - GVKHelp.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/GVKHelp.md))
    - ImportInspectionDialog.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/ImportInspectionDialog.md))
    - Installation.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/Installation.md))
    - ISBNtoBibTeXHelp.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/ISBNtoBibTeXHelp.md))
    - MedlineRIS.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/MedlineRIS.md))
    - MsOfficeBibFieldMapping.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/MsOfficeBibFieldMapping.md))
    - OpenOfficeIntegration.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/OpenOfficeIntegration.md))
    - RemoteHelp.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/RemoteHelp.md))
    - SharedBibFile.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/SharedBibFile.md))
    - SpecialFieldsHelp.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/SpecialFieldsHelp.md))
    - SpringerHelp.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/SpringerHelp.md))
    - SQLDatabase.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/SQLDatabase.md))
    - SQLDatabaseMigration.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/SQLDatabaseMigration.md))
    - SQLExport.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/SQLExport.md))
    - StringsHelp.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/StringsHelp.md))
    - XMPHelp.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/XMPHelp.md))
  - `ja`
    - ADSHelp.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/ADSHelp.md))
    - arXivHelp.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/arXivHelp.md))
    - BestPractices.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/BestPractices.md))
    - DBLPHelp.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/DBLPHelp.md))
    - DOAJHelp.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/DOAJHelp.md))
    - FindDuplicates.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/FindDuplicates.md))
    - GVKHelp.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/GVKHelp.md))
    - Installation.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/Installation.md))
    - MedlineRIS.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/MedlineRIS.md))
    - MsOfficeBibFieldMapping.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/MsOfficeBibFieldMapping.md))
    - SharedBibFile.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/SharedBibFile.md))
    - SpringerHelp.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/SpringerHelp.md))
    - SQLDatabase.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/SQLDatabase.md))
    - SQLDatabaseMigration.md ([en](https://github.com/JabRef/help.jabref.org/blob/gh-pages/en/SQLDatabaseMigration.md))

# running `update`
```
$ python _scripts/automate.py update
Language: 'de'
        redirected 25 page(s)
        0 page(s) with missing frontmatter
        created index with 56 page(s)
Language: 'en'
        0 page(s) with missing frontmatter
        created index with 56 page(s)
Language file 'fr' has an empty key 'Collaborative work'
Language file 'fr' has an empty key '... using online bibliographic database'
Language: 'fr'
        redirected 15 page(s)
        0 page(s) with missing frontmatter
        created index with 56 page(s)
Language file 'in' has an empty key 'You can't find a solution to your problem? You still have questions?'
Language file 'in' has an empty key 'Use the online forum to get more support!'
Language file 'in' has an empty key 'Collaborative work'
Language file 'in' has an empty key '... using publication identifiers'
Language file 'in' has an empty key '... using online bibliographic database'
Language: 'in'
        redirected 35 page(s)
        0 page(s) with missing frontmatter
        created index with 56 page(s)
Language file 'ja' has an empty key 'You can't find a solution to your problem? You still have questions?'
Language file 'ja' has an empty key 'Use the online forum to get more support!'
Language file 'ja' has an empty key 'Collaborative work'
Language file 'ja' has an empty key 'Fetching entries from the web'
Language file 'ja' has an empty key '... using publication identifiers'
Language file 'ja' has an empty key '... using online bibliographic database'
Language: 'ja'
        redirected 15 page(s)
        0 page(s) with missing frontmatter
        created index with 56 page(s)
```

# running `clean`
```
$ python _scripts/automate.py clean -e
Deleted 26 page(s) for language 'de'
        ['ACMPortalHelp.md', 'ADSHelp.md', 'arXivHelp.md', 'Autosave.md', 'BestPractices.md', 'DatabaseProperties.md', 'DBLPHelp.md', 'DiVAtoBibTeXHelp.md', 'DOAJHelp.md', 'DOItoBibTeXHelp.md', 'FindDuplicates.md', 'GoogleScholarHelp.md', 'GVKHelp.md', 'Installation.md', 'ISBNtoBibTeXHelp.md', 'LabelPatterns.md', 'MsOfficeBibFieldMapping.md', 'ScienceDirect.md', 'SharedBibFile.md', 'SpecialFieldsHelp.md', 'SpringerHelp.md', 'SQLDatabase.md', 'SQLDatabaseMigration.md', 'SQLExport.md', 'StringsHelp.md', 'index.md']
Deleted 1 page(s) for language 'en'
        ['index.md']
Deleted 16 page(s) for language 'fr'
        ['ADSHelp.md', 'arXivHelp.md', 'BestPractices.md', 'DBLPHelp.md', 'DOAJHelp.md', 'FindDuplicates.md', 'GVKHelp.md', 'Installation.md', 'LabelPatterns.md', 'MedlineRIS.md', 'MsOfficeBibFieldMapping.md', 'SharedBibFile.md', 'SpringerHelp.md', 'SQLDatabase.md', 'SQLDatabaseMigration.md', 'index.md']
Deleted 36 page(s) for language 'in'
        ['ADSHelp.md', 'arXivHelp.md', 'BestPractices.md', 'BibtexKeyPatterns.md', 'CommandLine.md', 'CustomEntriesHelp.md', 'CustomImports.md', 'DBLPHelp.md', 'DiVAtoBibTeXHelp.md', 'DOAJHelp.md', 'DOItoBibTeXHelp.md', 'EndNoteFilters.md', 'ExternalFiles.md', 'FileLinks.md', 'FindDuplicates.md', 'GeneralFields.md', 'GoogleScholarHelp.md', 'GroupsHelp.md', 'GVKHelp.md', 'ImportInspectionDialog.md', 'Installation.md', 'ISBNtoBibTeXHelp.md', 'LabelPatterns.md', 'MedlineRIS.md', 'MsOfficeBibFieldMapping.md', 'OpenOfficeIntegration.md', 'RemoteHelp.md', 'SharedBibFile.md', 'SpecialFieldsHelp.md', 'SpringerHelp.md', 'SQLDatabase.md', 'SQLDatabaseMigration.md', 'SQLExport.md', 'StringsHelp.md', 'XMPHelp.md', 'index.md']
Deleted 16 page(s) for language 'ja'
        ['ADSHelp.md', 'arXivHelp.md', 'BestPractices.md', 'DBLPHelp.md', 'DOAJHelp.md', 'FindDuplicates.md', 'GVKHelp.md', 'Installation.md', 'LabelPatterns.md', 'MedlineRIS.md', 'MsOfficeBibFieldMapping.md', 'SharedBibFile.md', 'SpringerHelp.md', 'SQLDatabase.md', 'SQLDatabaseMigration.md', 'index.md']
```

### CODE REVIEW

**Comment by Siedlerchr** [comment]
Awesome work! :heart: 
Just one little thing: Could you print the absolute number of pages somewhere, too, because having only the "relative" numbers, e.g. translated/non translated is a bit confusing right now on the first sight...I was wondering how the numbers for de added up until I saw the english column).

Maybe add en as reference in the first line?`

File: _scripts/automate.py [hunk 1]
```diff
@@ -0,0 +1,636 @@
+import argparse
+import codecs
+import datetime
+import json
+import os
+import subprocess
+import webbrowser
+from os import listdir
+from os.path import isfile, join, isdir
+
```
Comment (Siedlerchr) [inline _scripts/automate.py]: Python 2.x or 3.x?

**Comment by koppor** [comment]
Can you round the percent numbers to full percent? `39.2857142857` is really strange.

File: _scripts/automate.py (hunk 1 above)
Comment (bartsch-dev) [inline _scripts/automate.py]: it's Python 2.x

**Comment by bartsch-dev** [comment]
Now the table looks like that:

### Help pages status (2016-09-15 20:25 - Branch `addScriptToAutomateEverything` `bbda670`)
- Main language: `en`
- Available languages: `de`, `fr`, `in`, `ja`

| Language | translated | not translated | outdated | old pages | % translated | % outdated |
| -------- | ---------- | -------------- | -------- | --------- | ------------ | ---------- |
| en | 56 | 0 | 7 | 1 | 100 | 12 |
| de | 32 | 24 | 3 | 1 | 57 | 9 |
| fr | 42 | 14 | 3 | 1 | 75 | 7 |
| in | 22 | 34 | 1 | 1 | 39 | 4 |
| ja | 42 | 14 | 3 | 1 | 75 | 7 |

**Comment by koppor** [comment]
Could you please add support for removing the `Help` suffix (#19).

Two steps:

1. Only once:
`*Help.md` exists: Rename it to `*.md` and generate`*Help.md` redirecting to `*.md`.

2. In your script:
Ignore all files ending with `Help.md`.

**Comment by bartsch-dev** [comment]
Removed the Help suffix and added redirects for them.
All pages which only redirect were already ignored by my script.

**Comment by koppor** [comment]
@bartsch-dev Could you add a README.md into `_scripts` explaining the functionality? Maybe also a short note at README.md. I think CONTRIBUTING.md is the wrong place.

**Comment by koppor** [comment]
How can one generate the markdown table? The current help reads

```
usage: automate.py [-h] {status,update,clean,removeHelpSuffix} ...

positional arguments:
  {status,update,clean,removeHelpSuffix}
                        commands
    status              Lists the current status of the help pages.
    update              Generates all the missing redirection pages and
                        updates the index files
    clean               Removes all the generated redirect files (CAUTION:
                        index page may not work anymore
    removeHelpSuffix    Removes the 'Help' suffix from all pages and create
                        redirects

optional arguments:
  -h, --help            show this help message and exit
```

(Please also add the missing closing bracket at `clean`)

**Comment by bartsch-dev** [comment]
You can call the help on each of the positional arguments.
It's not perfect, maybe you can point me to a better arguments library.

`$ python _scripts/automate.py status -h`
```
usage: automate.py status [-h] [-e] [-m]

optional arguments:
  -h, --help      show this help message and exit
  -e, --extended  The output is much more sophisticated
  -m, --markdown  converts the output to markdown
```

**Pull request merged**